core/src/storage2/lazy/lazy_hmap.rs

```diff
@@ -296,8 +383,58 @@ where
     ///   with the underlying contract storage.
     /// - If the decoding of the old element at the given index failed.
     pub fn put(&mut self, key: K, new_value: Option<V>) {
-        self.entries_mut()
-            .insert(key, Box::new(Entry::new(new_value, EntryState::Mutated)));
+        self.entries_mut().insert(
+            key,
+            Box::new(StorageEntry::new(new_value, EntryState::Mutated)),
+        );
+    }
+}
+
+impl<K, V, H> LazyHashMap<K, V, H>
+where
+    K: Clone + Ord + PackedLayout,
+    V: PackedLayout,
+    H: Hasher,
+    Key: From<<H as Hasher>::Output>,
+{
+    /// Gets the given key's corresponding entry in the map for in-place manipulation.
+    pub fn entry(&mut self, key: K) -> Entry<K, V> {
+        // SAFETY: We have put the whole `cached_entries` mapping into an
+        //         `UnsafeCell` because of this caching functionality. The
+        //         trick here is that due to using `Box<T>` internally
+        //         we are able to return references to the cached entries
+        //         while maintaining the invariant that mutating the caching
+        //         `BTreeMap` will never invalidate those references.
+        //         By returning a raw pointer we enforce an `unsafe` block at
+        //         the caller site to underline that guarantees are given by the
+        //         caller.
+        let cached_entries = unsafe { &mut *self.cached_entries.get_ptr().as_ptr() };
+        // We have to clone the key here because we do not have access to the unsafe
+        // raw entry API for Rust hash maps, yet since it is unstable. We can remove
+        // the constraints on `K: Clone` once we have access to this API.
+        // Read more about the issue here: https://github.com/rust-lang/rust/issues/56167
+        match cached_entries.entry(key.to_owned()) {
+            BTreeMapEntry::Occupied(entry) => {
+                match entry.get().value() {
+                    Some(_) => Entry::Occupied(OccupiedEntry { key, entry }),
+                    None => {
+                        // value is already marked as to be removed
+                        Entry::Vacant(VacantEntry {
+                            key,
+                            entry: BTreeMapEntry::Occupied(entry),
+                        })
+                    }
+                }
+            }
+            BTreeMapEntry::Vacant(vacant) => {
+                let value = self
+                    .key_at(&key)
+                    .map(|key| pull_packed_root_opt::<V>(&key))
+                    .unwrap_or(None);
+                vacant.insert(Box::new(StorageEntry::new(value, EntryState::Preserved)));
+                self.entry(key)
```

Comment:
If you call this method recursively you won again nothing in the benchmarks since you query the underlying caching b-tree map multiple times now which is exactly what the entry API tries to avoid. So you have to find another method in order to return your entry properly without querying the underlying b-tree map multiple times just as you did in the occupied state.